org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/semanticHighlight/SemanticHighlightingRegistry.xtend

```diff
@@ -0,0 +1,221 @@
+/*******************************************************************************
+ * Copyright (c) 2018 TypeFox and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.ide.server.semanticHighlight
+
+import com.google.common.base.Preconditions
+import com.google.common.collect.BiMap
+import com.google.common.collect.ImmutableBiMap
+import com.google.common.collect.ImmutableList
+import com.google.common.collect.ImmutableMultimap
+import com.google.inject.Inject
+import com.google.inject.Singleton
+import java.util.List
+import org.eclipse.lsp4j.ClientCapabilities
+import org.eclipse.lsp4j.Position
+import org.eclipse.lsp4j.Range
+import org.eclipse.lsp4j.SemanticHighlightingInformation
+import org.eclipse.lsp4j.SemanticHighlightingParams
+import org.eclipse.lsp4j.VersionedTextDocumentIdentifier
+import org.eclipse.lsp4j.services.LanguageClient
+import org.eclipse.lsp4j.util.SemanticHighlightingTokens
+import org.eclipse.xtend.lib.annotations.Data
+import org.eclipse.xtext.ide.editor.syntaxcoloring.ISemanticHighlightingCalculator
+import org.eclipse.xtext.ide.editor.syntaxcoloring.MergingHighlightedPositionAcceptor
+import org.eclipse.xtext.ide.server.Document
+import org.eclipse.xtext.ide.server.ILanguageServerAccess
+import org.eclipse.xtext.ide.server.UriExtensions
+import org.eclipse.xtext.resource.IResourceServiceProvider
+import org.eclipse.xtext.resource.XtextResource
+import org.eclipse.xtext.util.CancelIndicator
+import org.eclipse.xtext.util.internal.Log
+
+import static extension org.eclipse.lsp4j.util.SemanticHighlightingTokens.encode
+
+/**
+ * Shared semantic highlighting manager per language server.
+ * Responsible for converting the semantic highlighted ranges into the LSP standard by producing a compact,
+ * {@code base64} encoded token string.
+ * 
+ */
+@Log
+@Singleton
```

Comment:
It must not be `@Singleton`. Otherwise, the initialize method will throw an ISE when the client tries to reconnect. I am verifying this locally and will update the PR according to it.

Reply:
Yes, it is a bug.